pkg/binlog/event/generator.go

```diff
@@ -0,0 +1,134 @@
+		ServerID:      serverID,
+		LatestPos:     latestPos,
+		LatestGTID:    latestGTID,
+		PreviousGTIDs: previousGTIDs,
+		LatestXID:     latestXID,
+	}
+}
+
+// GenFileHeader generates a binlog file header, including to PreviousGTIDsEvent/MariadbGTIDListEvent.
+// for MySQL:
+//   1. BinLogFileHeader, [ fe `bin` ]
+//   2. FormatDescriptionEvent
+//   3. PreviousGTIDsEvent
+// for MariaDB:
+//   1. BinLogFileHeader, [ fe `bin` ]
+//   2. FormatDescriptionEvent
+//   3. MariadbGTIDListEvent
+func (g *Generator) GenFileHeader() ([]*replication.BinlogEvent, []byte, error) {
+	events, data, err := GenCommonFileHeader(g.Flavor, g.ServerID, g.PreviousGTIDs)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.LatestPos = uint32(len(data)) // if generate a binlog file header then reset latest pos
+	return events, data, nil
+}
+
+// GenCreateDatabaseEvents generates binlog events for `CREATE DATABASE`.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenCreateDatabaseEvents(schema string) ([]*replication.BinlogEvent, []byte, error) {
+	result, err := GenCreateDatabaseEvents(g.Flavor, g.ServerID, g.LatestPos, g.LatestGTID, schema)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.updateLatestPosGTID(result.LatestPos, result.LatestGTID)
+	return result.Events, result.Data, nil
+}
+
+// GenDropDatabaseEvents generates binlog events for `DROP DATABASE`.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenDropDatabaseEvents(schema string) ([]*replication.BinlogEvent, []byte, error) {
+	result, err := GenDropDatabaseEvents(g.Flavor, g.ServerID, g.LatestPos, g.LatestGTID, schema)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.updateLatestPosGTID(result.LatestPos, result.LatestGTID)
+	return result.Events, result.Data, nil
+}
+
+// GenCreateTableEvents generates binlog events for `CREATE TABLE`.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenCreateTableEvents(schema string, query string) ([]*replication.BinlogEvent, []byte, error) {
+	result, err := GenCreateTableEvents(g.Flavor, g.ServerID, g.LatestPos, g.LatestGTID, schema, query)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.updateLatestPosGTID(result.LatestPos, result.LatestGTID)
+	return result.Events, result.Data, nil
+}
+
+// GenDropTableEvents generates binlog events for `DROP TABLE`.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenDropTableEvents(schema string, table string) ([]*replication.BinlogEvent, []byte, error) {
+	result, err := GenDropTableEvents(g.Flavor, g.ServerID, g.LatestPos, g.LatestGTID, schema, table)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.updateLatestPosGTID(result.LatestPos, result.LatestGTID)
+	return result.Events, result.Data, nil
+}
+
+// GenDDLEvents generates binlog events for DDL statements.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenDDLEvents(schema string, query string) ([]*replication.BinlogEvent, []byte, error) {
```

Comment:
I perfer it

Reply:
yeh. In fact, only this is enough for DDL.